### ISSUE TITLE
Improved Physics Support

### ISSUE THREAD

**Issue by vlidholt** [spritebuilder/SpriteBuilder#224]
Main new feature is support for visually editing joints. This issue needs to be fleshed out before starting working on it.

Key things to take into consideration:
- Add the joints to the top-level inside a ccb-file. It needs to be possible to construct complete physics objects in sub nodes. E.g. a car that you can drag into your main scene.
- How to reference the objects from the joints inside the ccb and ccbi file. Will it be needed to add some sort of identifiers to the objects that are using joints? If so, it needs to be done in a way that can be reused by other potential future features.
- How should the user interface look for this feature? Should the different joints be available in the plug-in view, or should we create a new tab for them? How do we visually connect the joints to the actual objects? Need to update the display when moving things around, etc.

**Comment by vlidholt** [comment]
Starting to take shape! :) Pivot joints are already somewhat working.

For the joints (specifically the pivot joints):

DONE- The joints doesn’t use the selected images when selected
DONE - It’s possible to drag the joints in the hierarchy to invalid places (which breaks things) on the Sequncer View
DONE - When moving objects with physics bodies around in the hierarchy view the physics joint breaks (sometimes you see it breaking, sometimes it just stops working) on the Hierarchy. eg: moving a body around, because SB deletes and creates a node. Todo: Only story UUID, and Get CCnode each time.
DONE - Make it possible to hook up nodes from the property inspector and dragging to the node hierarchy. This likely requires a transparent window on top of the main window during the time of the drag operation. Added benefit is that we could do nicer drawing of the connection line with antialiasing.

DONE - Shape references in the inspector needs to be read-only.
- Remove connection (X) and 'Select connected nodes'-button nicer icons. 'Select connected node'-button should have similar look to the connection on the stage.

DONE - Dragging a joint to the timeline causes a crash.

DONE - Disabling physics for a body that is connected to a joint doesn't invalidate the joint connection.

DONE - Need to be able to set more options for the joints: collideBodies, maxForce, breakingForce. These should probably go above the other properties in the inspector under a 'CCPhysicsJoint' section. Followed by the 'CCPhysicsJointPivot' section.

DONE - SpriteBuilder crashes when clicking the fold-out button on the 'Joints' track and on the 'CCPhysicsPivotJoint' track. There should be no fold-out button.

DONE - Disable the tabs on the right hand side for physics bodies if a joint is selected or if a CCPhysicsNode is selected (add an extra property to the plug-ins for this).

DONE - Add more properties: sensor, collisionGroup, collisionCategories, collisionMask

**Comment by jtwigg** [comment]
DONE -Cannot drag from joint to timeline
DONE -Cannot drag from outlet to 'Circular body shape' bodies.
DONE -Cutting and pasting a body doesn't repair the joint reference.

**Comment by jtwigg** [comment]
- Copy and paste and clone joints.
- Copy and paste across documents.

**Comment by jtwigg** [comment]
If physics body is enabled, disable Skew.

**Comment by vlidholt** [comment]
Did some testing today, these are mine findings:

DONE - Selecting the SequenecerJoint node visually on the canvas doesn’t select it in inspector and doesn’t select it in the time-line
DONE - "Collide bodies" check doesn't work cannot be disabled (should probably be disabled by default)
- Sometimes it messes up BodyA/B and attaches incorrect nodes (maybe because I have a sub ccb-file with physics nodes in it?)
- The publishing phase sometimes complains incorrectly about missing BodyA/B (not consistently, I had it happen approx. 1/3 of times I published)

**Comment by jtwigg** [comment]
DONE: Collide bodies – should be false by default (as it’s likely to be the most common user case)

DONE : Collision categories & masks – use a NSTokenField for this (currently not intuitively clear how to separate multiple types, commas? spaces?)

DONE : Gray out the physics tab for joints & CCPhysicsNode (and if there are other places where it doesn’t make sense to have it enabled)

DONE : Pin joints should be called distance joints (as they are in CCPhysics)

DONE : I think we are better off joining the plug-ins for the joints into their own files. For instance, a 
CCPhysicsSpringJoint is not a sub-class of CCPhysicsPivotJoint, also it makes sense from a user point of view to have the “Anchor A” and “Anchor B” grouped together.

DONE For the breaking/max force see the UI mockup below, I think it would make a bit more sense as it does’t have different states that produces the same value.

DONE : Attached new gfx for the distance/spring joints. It adds a pixels padding so that anti-alias will look better. I changed the image slightly so the end caps are now 25% of the image. You may also need to do some pixel adjustments depending on how your code look.

DONE : Adding graphics for body connections (see below and attached zip)

DONE : Spring rest length doesn’t work correctly (seems to be offset in relation to the mouse pointer, also can be dragged to a negative value & highlight is off sometimes before mouse over)

DONE : I think that the currently selected joint needs to be drawn on top (maybe just set the z order?), as they are otherwise hard to edit if other joints overlap

Done: Selecting multiple joints renders unexpected visual results (maybe it should only be possible to select one at a time?)

**Comment by jtwigg** [comment]
DONE: Undo relocates a joint to a center point.

**Comment by vlidholt** [comment]
Posting issues here as I find them:
- [x] Joints connection points needs to be drawn aligned to pixels, now they get blurry when dragged around.
- [x] I can only move a pivot joint by selecting it’s very center point. Feels like I should be able to move the joint around by dragging the larger circle. (No big deal though, it’s quick to figure out)
- [x] When dragging and dropping a sub-ccb file I can connect bodies the sub-file bodies in the super file, but the joints have no effect in the run-time.
- [x] The inspector has moved down a pixel? Somehow the top line is too thick (compare other tabs & see pic below)
- [x] Use a reasonable default value for breaking force and max force if infinity isn’t selected (default value shouldn’t be infinity).
- [ ] The mouse over effect when dragging the handles on a joint look very aliased, would look better without the hoover effect (or we have to draw something better for this).
- [x] If you drag a pivot joint to the stage and then connect it's left connector the pivot joint jumps (the more intuitive behavior would be if the joint stayed in place).
- [x] Body names can be edited, but things break when you do it.
- [x] The goto arrow shouldn't be visible when there is no body connected to a joint
- [x] Disconnect button should be grayed & disabled out when no node is connected to a joint
- [x] There should be a option for hiding physics in the Document menu. Selecting a joint in the timeline, clicking the physics tab, etc should automatically toggle on physics display
- [x] Inspector property types needs to be aligned an constantly named (use of capital letters etc). _Vik_
- [x] New icons for node library view. _Vik_
- [x] Shouldnt be able to attach a physics body to a sub-ccb file.
  Line thickness in inspector:
  ![line thickness](https://cloud.githubusercontent.com/assets/1539812/2701424/e971b598-c420-11e3-8426-52bc327b154b.png)

**Comment by jtwigg** [comment]
- [x] Add enable switch to Rest Length
- [x] Visible toggle breaks when animation is played.
- [x] When cloning timelines, it duplicates autoPlay.

**Comment by vlidholt** [comment]
A couple of more things:
- [x] Selections looks funky when more than one physics joint is selected
- [x] Dragging more than one joint on the stage works weirdly (it's ok to disable dragging of multiple joints)
- [x] Selecting a joint in the timeline or dragging a joint on stage doesn't turn on the visibility of joints
- [x] Perhaps the 'show joints' in Document menu should toggle the visibility icon in the timeline, currently these features duplicate the behavior which may be confusing to users (you can hide in two different places and both needs to be turned on to actually see the joints)

**Comment by vlidholt** [comment]
- [x] For breaking force and max force, it looks like you have the options between infinity or infinity. It should instead display the default value in the grayed out field. (The value that is set when you select the 'other' infinity). See A
- [x] From Minimum & maximum distance, it is by default displayed as infinity (positive or negative), but when you enable it the value changes. By disabling this field it doesn't return to it's initial value. Best would be if the default values were always displayed. Second best, leaving the field blank when it's not enabled. See B

![forces](https://cloud.githubusercontent.com/assets/1539812/2760024/d845c83c-c9ab-11e3-8076-fab9e6b745c9.png)

**Comment by jtwigg** [comment]
- [x] Dragging mouse sequencer window over joints has problems
- [x] Ctrl Clicking on joints causes exception.

**Comment by jtwigg** [comment]
@vlidholt  Can we close this?

**Comment by vlidholt** [comment]
Closing, found one issue, but placing it in a separate issue.
